File: ts/src/http.ts
import { TokenManager } from "./auth/token-manager.js";
import { AuthError, ApiError, NetworkError } from "./errors.js";

/** Detect if a token is a JWT (3 dot-separated base64 parts, starts with eyJ).
 *  Legacy opaque tokens like 'login_acc_...' always return false.
 */
function isJwt(token: string): boolean {
  const parts = token.split(".");
  return parts.length === 3 && token.startsWith("eyJ");
}

interface TransportOptions {
  apiKey?: string;
  timeout: number;
  tokenManager: TokenManager;
  organizationId?: string;
}

export class HttpTransport {
  constructor(private readonly opts: TransportOptions) {}

  public clearApiKey(): void {
    this.opts.apiKey = undefined
  }

  private async sleep(ms: number) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  public getFetch(): typeof fetch {
    return async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
      let retries = 0;
      const maxRetries = 2;

      while (true) {
        const controller = new AbortController();
        const parentSignal = init?.signal;
        
        let timeoutId: ReturnType<typeof setTimeout> | undefined;
        
        if (parentSignal?.aborted) {
          throw new NetworkError("Request aborted by user");
        }

        const abortListener = () => controller.abort();
        if (parentSignal) {
          parentSignal.addEventListener('abort', abortListener);
        }

        timeoutId = setTimeout(() => controller.abort(), this.opts.timeout);
        
        const reqInit: RequestInit = {
          ...init,
          signal: controller.signal,
          headers: new Headers(init?.headers),
        };

        const headers = reqInit.headers as Headers;
        const token = this.opts.tokenManager.getToken();

        if (this.opts.apiKey) {
          // API Key mode: server-to-server, org resolved by gateway from key
          headers.set("x-api-key", this.opts.apiKey);
        } else if (token && this.opts.organizationId && isJwt(token)) {
          // JWT Bearer mode: auth-service JWT + org scope (only for real JWTs)
          headers.set("authorization", `Bearer ${token}`);
          headers.set("x-organization-id", this.opts.organizationId);
        } else if (token) {
          // Legacy access token mode (opaque tokens like login_acc_...)
          headers.set("x-access-token", token);
        }

        try {
          const res = await globalThis.fetch(input, reqInit);
          
          if (timeoutId) clearTimeout(timeoutId);
          if (parentSignal) parentSignal.removeEventListener('abort', abortListener);

          if (res.ok) {
            return res;
          }

          if (res.status === 401 || res.status === 403) {
            if (this.opts.apiKey) {
              throw new AuthError("Invalid or expired API key (x-api-key).");
            } else if (token && this.opts.organizationId && isJwt(token)) {
              throw new AuthError("Invalid or expired JWT token (Authorization: Bearer) or organizationId not in token.");
            } else if (token) {
              throw new AuthError("Invalid or expired access token (x-access-token).");
            }
            throw new AuthError("No credentials provided — set accessToken= (user login) or apiKey= (server-to-server).");
          }

          if ((res.status === 429 || res.status >= 500) && retries < maxRetries) {
            retries++;
            await this.sleep(1000 * Math.pow(2, retries));
            continue;
          }

          const text = await res.text().catch(() => "Unknown error");
          throw new ApiError(res.status, text);

        } catch (error) {
          if (timeoutId) clearTimeout(timeoutId);
          if (parentSignal) parentSignal.removeEventListener('abort', abortListener);

          if (error instanceof AuthError || error instanceof ApiError) {
            throw error;
          }
          
          if (error instanceof Error && error.name === 'AbortError') {
             if (parentSignal?.aborted) {
               throw new NetworkError("Request aborted by user");
             }
             
             if (retries < maxRetries) {
               retries++;
               await this.sleep(1000 * Math.pow(2, retries));
               continue;
             }
             throw new NetworkError(`Request timed out after ${this.opts.timeout}ms`);
          }

          if (retries < maxRetries) {
            retries++;
            await this.sleep(1000 * Math.pow(2, retries));
            continue;
          }

          throw new NetworkError(error instanceof Error ? error.message : "Network error or server unreachable");
        }
      }
    };
  }
}

File: py/src/imbrace/http.py
import time
import asyncio
import logging
from typing import Any, AsyncIterator, Dict, Iterator, List, Optional, Type, TypeVar
import httpx
from pydantic import BaseModel

from .auth.token_manager import TokenManager
from .exceptions import AuthError, ApiError, NetworkError

logger = logging.getLogger("imbrace")
T = TypeVar("T", bound=BaseModel)

def _is_jwt(token: str) -> bool:
    """Detect if a token is a JWT (3 dot-separated base64 parts, starts with eyJ).
    Legacy opaque tokens like 'login_acc_...' always return False.
    """
    parts = token.split(".")
    return len(parts) == 3 and token.startswith("eyJ")


class HttpTransport:
    """Synchronous HTTP Transport Layer for Imbrace SDK."""
    def __init__(
        self,
        token_manager: TokenManager,
        timeout: int = 30,
        api_key: Optional[str] = None,
        organization_id: Optional[str] = None,
    ):
        self.token_manager = token_manager
        self.timeout = timeout
        self.api_key = api_key
        self.organization_id = organization_id
        self._client = httpx.Client(
            timeout=timeout,
            limits=httpx.Limits(max_keepalive_connections=5, max_connections=10)
        )
        logger.debug("Initialized HttpTransport (sync)")

    def _auth_error(self, token: Optional[str]) -> AuthError:
        if self.api_key:
            return AuthError("Invalid or expired API key (x-api-key).")
        if token and self.organization_id and _is_jwt(token):
            return AuthError("Invalid or expired JWT token (Authorization: Bearer) or organization_id not in token.")
        if token:
            return AuthError("Invalid or expired access token (x-access-token).")
        return AuthError("No credentials provided — set access_token= (user login) or api_key= (server-to-server).")

    def request(self, method: str, url: str, **kwargs) -> httpx.Response:
        retries = 0
        max_retries = 3

        headers = kwargs.pop("headers", {})
        token = self.token_manager.get_token()

        if self.api_key:
            # API Key mode: server-to-server, org resolved by gateway from key
            headers["x-api-key"] = self.api_key
        elif token and self.organization_id and _is_jwt(token):
            # JWT Bearer mode: auth-service JWT + org scope (only for real JWTs)
            headers["authorization"] = f"Bearer {token}"
            headers["x-organization-id"] = self.organization_id
        elif token:
            # Legacy access token mode (opaque tokens like login_acc_...)
            headers["x-access-token"] = token

        kwargs["headers"] = headers

        while True:
            try:
                logger.debug(f"Request: {method} {url}")
                res = self._client.request(method, url, **kwargs)
                
                if res.status_code < 400:
                    return res
                if res.status_code in [401, 403]:
                    logger.error(f"Auth error {res.status_code} at {url}")
                    raise self._auth_error(token)
                if (res.status_code == 429 or res.status_code >= 500) and retries < max_retries:
                    wait_time = 2 ** retries
                    logger.warning(f"Retry {retries + 1}/{max_retries} after {wait_time}s due to status {res.status_code}")
                    retries += 1
                    time.sleep(wait_time)
                    continue
                    
                logger.error(f"API Error {res.status_code}: {res.text}")
                raise ApiError(res.status_code, res.text)
            except httpx.RequestError as e:
                if retries < max_retries:
                    wait_time = 2 ** retries
                    logger.warning(f"Retry {retries + 1}/{max_retries} after {wait_time}s due to network error: {str(e)}")
                    retries += 1
                    time.sleep(wait_time)
                    continue
                raise NetworkError(f"Network error or timeout: {str(e)}")

    def iterate_paged(
        self, 
        method: str, 
        url: str, 
        model: Type[T], 
        params: Optional[Dict[str, Any]] = None,
        **kwargs
    ) -> Iterator[T]:
        """Tự động duyệt qua tất cả các trang của một API phân trang."""
        current_params = (params or {}).copy()
        if "page" not in current_params:
            current_params["page"] = 1
        
        while True:
            res = self.request(method, url, params=current_params, **kwargs).json()
            # Giả định cấu hình PagedResponse: { data: [...], pagination: { has_next: bool } }
            data = res.get("data", [])
            for item in data:
                yield model(**item)
            
            pagination = res.get("pagination", {})
            if not pagination.get("has_next"):
                break
            current_params["page"] += 1

    def close(self):
        self._client.close()


class AsyncHttpTransport:
    """Asynchronous HTTP Transport Layer for Imbrace SDK."""
    def __init__(
        self,
        token_manager: TokenManager,
        timeout: int = 30,
        api_key: Optional[str] = None,
        organization_id: Optional[str] = None,
    ):
        self.token_manager = token_manager
        self.timeout = timeout
        self.api_key = api_key
        self.organization_id = organization_id
        self._client = httpx.AsyncClient(
            timeout=timeout,
            limits=httpx.Limits(max_keepalive_connections=5, max_connections=10)
        )
        logger.debug("Initialized AsyncHttpTransport")

    def _auth_error(self, token: Optional[str]) -> AuthError:
        if self.api_key:
            return AuthError("Invalid or expired API key (x-api-key).")
        if token and self.organization_id and _is_jwt(token):
            return AuthError("Invalid or expired JWT token (Authorization: Bearer) or organization_id not in token.")
        if token:
            return AuthError("Invalid or expired access token (x-access-token).")
        return AuthError("No credentials provided — set access_token= (user login) or api_key= (server-to-server).")

    async def request(self, method: str, url: str, **kwargs) -> httpx.Response:
        retries = 0
        max_retries = 3

        headers = kwargs.pop("headers", {})
        token = self.token_manager.get_token()

        if self.api_key:
            # API Key mode: server-to-server, org resolved by gateway from key
            headers["x-api-key"] = self.api_key
        elif token and self.organization_id and _is_jwt(token):
            # JWT Bearer mode: auth-service JWT + org scope (only for real JWTs)
            headers["authorization"] = f"Bearer {token}"
            headers["x-organization-id"] = self.organization_id
        elif token:
            # Legacy access token mode (opaque tokens like login_acc_...)
            headers["x-access-token"] = token

        kwargs["headers"] = headers

        while True:
            try:
                res = await self._client.request(method, url, **kwargs)
                if res.status_code < 400:
                    return res
                if res.status_code in [401, 403]:
                    raise self._auth_error(token)
                if (res.status_code == 429 or res.status_code >= 500) and retries < max_retries:
                    wait_time = 2 ** retries
                    retries += 1
                    await asyncio.sleep(wait_time)
                    continue
                raise ApiError(res.status_code, res.text)
            except httpx.RequestError as e:
                if retries < max_retries:
                    wait_time = 2 ** retries
                    retries += 1
                    await asyncio.sleep(wait_time)
                    continue
                raise NetworkError(f"Network error or timeout: {str(e)}")

    async def iterate_paged(
        self, 
        method: str, 
        url: str, 
        model: Type[T], 
        params: Optional[Dict[str, Any]] = None,
        **kwargs
    ) -> AsyncIterator[T]:
        """Tự động duyệt qua tất cả các trang (bất đồng bộ)."""
        current_params = (params or {}).copy()
        if "page" not in current_params:
            current_params["page"] = 1
        
        while True:
            res = await self.request(method, url, params=current_params, **kwargs)
            body = res.json()
            data = body.get("data", [])
            for item in data:
                yield model(**item)
            
            pagination = body.get("pagination", {})
            if not pagination.get("has_next"):
                break
            current_params["page"] += 1

    async def close(self):
        await self._client.aclose()
